src/api/trace/b3_propagator.rs

```diff
@@ -3,62 +3,114 @@
 //! The `B3Propagator` facilitates `SpanContext` propagation using
 //! B3 Headers. This propagator supports both version of B3 headers,
 //!  1. Single Header:
-//!    X-B3: {trace_id}-{span_id}-{sampling_state}-{parent_span_id}
+//!    b3: {trace_id}-{span_id}-{sampling_state}-{parent_span_id}
 //!  2. Multiple Headers:
 //!    X-B3-TraceId: {trace_id}
 //!    X-B3-ParentSpanId: {parent_span_id}
 //!    X-B3-SpanId: {span_id}
 //!    X-B3-Sampled: {sampling_state}
 //!    X-B3-Flags: {debug_flag}
 //!
-//! If `single_header` is set to `true` then `X-B3` header is used to inject
+//! If `inject_encoding` is set to `B3Encoding::B3SingleHeader` then `b3` header is used to inject
 //! and extract. Otherwise, separate headers are used to inject and extract.
 use crate::{api, api::TraceContextExt};
+use crate::api::TRACE_FLAG_DEFERRED;
+
+static B3_SINGLE_HEADER: &str = "b3";
+/// As per spec, the multiple header should be case sensitive. But different protocol will use
+/// different format. For example, HTTP will use X-B3-$name while gRPC will use x-b3-$name. So here
+/// we leave it to be lower case since we cannot tell what kind of protocol will be used.
+/// Go implementation also uses lower case.
+static B3_DEBUG_FLAG_HEADER: &str = "x-b3-flags";
+static B3_TRACE_ID_HEADER: &str = "x-b3-traceid";
+static B3_SPAN_ID_HEADER: &str = "x-b3-spanid";
+static B3_SAMPLED_HEADER: &str = "x-b3-sampled";
+static B3_PARENT_SPAN_ID_HEADER: &str = "x-b3-parentspanid";
+
+/// B3Encoding is a bitmask to represent B3 encoding type
+#[derive(Clone, Debug, Copy)]
+pub enum B3Encoding {
+    /// B3Unspecified is an unspecified B3 encoding
+    B3Unspecified = 0,
+    /// B3MultipleHeader is a B3 encoding that uses multiple headers
+    /// to transmit tracing information prefixed with `X-B3-`
+    B3MultipleHeader = 1,
+    /// B3SingleHeader is B3 encoding that uses a single header named `b3`
+    /// to transmit tracing information
+    B3SingleHeader = 2,
+    /// B3SingleAndMultiHeader is B3 encoding that uses both single header and multiple headers
+    /// to transmit tracing information. Note that if both single header and multiple headers are
+    /// provided, the single header will take precedence when extracted
+    B3SingleAndMultiHeader = 3,
+}
 
-static B3_SINGLE_HEADER: &str = "X-B3";
-static B3_DEBUG_FLAG_HEADER: &str = "X-B3-Flags";
-static B3_TRACE_ID_HEADER: &str = "X-B3-TraceId";
-static B3_SPAN_ID_HEADER: &str = "X-B3-SpanId";
-static B3_SAMPLED_HEADER: &str = "X-B3-Sampled";
-static B3_PARENT_SPAN_ID_HEADER: &str = "X-B3-ParentSpanId";
+impl B3Encoding {
+    /// support determines if current encoding supports the `e`
+    pub fn support(&self, e: B3Encoding) -> bool {
```

Comment:
small nit: could have e be `other: &Self` here to signal that it is comparing same types? Not necessary but seems more standard sometimes for rust APIs.